org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/JavaCodeMiningReconciler.java

```diff
@@ -47,15 +48,22 @@ public class JavaCodeMiningReconciler implements IJavaReconcilingListener {
 	public void reconciled(CompilationUnit ast, boolean forced, IProgressMonitor progressMonitor) {
 		final ISourceViewerExtension5 sourceViewer= fSourceViewer; // take a copy as this can be null-ed in the meantime
 		if (sourceViewer != null) {
-			reconciledViewers.add(sourceViewer);
+			CompletableFuture<CompilationUnit> future= toBeReconciledViewers.remove(sourceViewer);
+			if (future!=null) {
+				future.complete(ast);
+			}
 			sourceViewer.updateCodeMinings();
 		}
 	}
 
+	public static CompletableFuture<CompilationUnit> getFuture(ISourceViewerExtension5 viewer) {
+		CompletableFuture<CompilationUnit> future= toBeReconciledViewers.get(viewer);
+		return future;
```

Comment:
You should use computeIfAbsent to make sure you always get one, even if I assume due to the code flow it should probably not be needed. This also should include checking if we already have some data (is that possible?) and then retrun an already completed future.

Reply:
This is not a good design. If the source is reconciled it should here return an already completed future with the result, and the client should only ever call this method and there is no need to distinguish the different cases.